xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -0,0 +1,313 @@
+        """
+
+        # this is the unbufferized `tensor<(neighbours)x(ZDim)x(type)>` value
+        assert isa(typ := op.chunk_reduce.block.args[0].type, TensorType[Attribute])
+
+        return tensor.ExtractSliceOp(
+            operands=[to_tensor.tensor, [offset], [], []],
+            result_types=[TensorType(typ.get_element_type(), typ.get_shape()[1:])],
+            properties={
+                "static_offsets": DenseArrayBase.from_list(
+                    i64, (memref.Subview.DYNAMIC_INDEX,)
+                ),
+                "static_sizes": DenseArrayBase.from_list(i64, typ.get_shape()[1:]),
+                "static_strides": DenseArrayBase.from_list(i64, (1,)),
+            },
+        )
+
+
+@dataclass(frozen=True)
+class AccessOpBufferize(RewritePattern):
+    """Bufferizes AccessOp."""
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.AccessOp, rewriter: PatternRewriter, /):
+        if not isa(op.result.type, TensorType[Attribute]):
+            return
+        rewriter.replace_matched_op(
+            [
+                access := csl_stencil.AccessOp(
+                    op.op,
+                    op.offset,
+                    tensor_to_memref_type(op.result.type),
+                    op.offset_mapping,
+                ),
+                to_tensor_op(access.result),
+            ]
+        )
+
+
+@dataclass(frozen=True)
+class YieldOpBufferize(RewritePattern):
+    """Bufferizes YieldOp."""
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.YieldOp, rewriter: PatternRewriter, /):
+        to_memrefs: list[Operation] = []
+        args: list[SSAValue] = []
+        for arg in op.arguments:
+            if isa(arg.type, TensorType[Attribute]):
+                to_memrefs.append(new_arg := to_memref_op(arg))
+                args.append(new_arg.memref)
+            else:
+                args.append(arg)
+
+        if len(to_memrefs) == 0:
+            return
+
+        rewriter.replace_matched_op([*to_memrefs, csl_stencil.YieldOp(*args)])
+
+
+@dataclass(frozen=True)
+class FuncOpBufferize(RewritePattern):
+    """
+    Replace the function_type and let a separate type conversion pass handle the block args.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
+        op.function_type = FunctionType.from_lists(
+            [
+                (
+                    tensor_to_memref_type(t.get_element_type())
+                    if isa(t, stencil.FieldType[TensorType[Attribute]])
+                    else t
+                )
+                for t in op.function_type.inputs
+            ],
+            [
+                (
+                    tensor_to_memref_type(t.get_element_type())
+                    if isa(t, stencil.FieldType[TensorType[Attribute]])
+                    else t
+                )
+                for t in op.function_type.outputs
+            ],
+        )
```

Comment:
I did that in a few places but it would be advisable to replace with a new op having this new result type;
For things like proper worklist management and all; doing it this way, you might get surprises if you expect it to trigger other patterns!
I don't mind myself about merging it either way; just a note

Reply:
Thanks good point, done as suggested.